topology changes to enable Node-Local Services Access in Shared GW Mode

girishmg opened the pull request:
In the shared gateway mode, the current OVN Logical Topology does not
support hostNetwork pods from accessing a kubernetes ClusterIP service
endpoint which happens to be the node itself. Such traffic gets dropped.

In this mode, we share the IP address of the node to SNAT the traffic
going out of the OVN Logical network to the physical network. The OVN
Gateway router does this SNATing. When a packet with destination IP set
to the Node IP arrives on the `ingress` pipeline of the Gateway Router,
it expects the packet to be one of these things:
  (1) a packet for NodePort service and it DNATs it to one of the
     backend IPs
  (2) a reply packet for an already established (or in the process of)
     connection
If it sees a packet from the OVN logical topology towards the Node IP,
then it gets dropped since it doesn’t fit (1) or (2) above.

This commit provides a path for such traffic by using OVN features of
Policy-Based routing and one-to-one DNAT and SNAT on the distributed
gateway port.

The google doc version of the design document is here:
https://docs.google.com/document/d/1DfbcAezbZD4Ipswz6UOn30S6B6f6eoiLkTCVmTa43Vk/edit?usp=sharing

The Markup version of the design document along with the diagrams is in
docs/design/shared_gw_dgp.md

girishmg commented:
@ovn-org/ovn-kubernetes-committers This PR is ready for review. PTAL. The design document has detailed information on the incorporated changes.

trozet commented on go-controller/pkg/ovn/ipallocator/allocator.go:
```diff
@@ -0,0 +1,231 @@
+package ipallocator
```
This code is all copied from https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/service/ipallocator/allocator.go ?

girishmg replied:
Right. I have submitted a separate PR for this. Please see #1374

trozet commented on go-controller/pkg/ovn/ipallocator/bitmap.go:
```diff
@@ -0,0 +1,197 @@
+package ipallocator
```
copied from https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/service/allocator/bitmap.go ? If so maybe add a comment stating that and that we dont need to unit test it? Or should you also copy in their unit test so our coverage doesn't fall?

girishmg replied:
In the new PR, I have included the unit tests as well. Please see PR #1374

trozet commented on go-controller/pkg/ovn/master.go:
```diff
@@ -710,7 +695,7 @@ func (oc *Controller) deleteNodeLogicalNetwork(nodeName string) error {
 }
 
-func (oc *Controller) deleteNode(nodeName string, hostSubnets, joinSubnets []*net.IPNet) error {
+func (oc *Controller) deleteNode(nodeName string, hostSubnets, joinSubnets []*net.IPNet, dnatSnatIP []net.IP) error {
```
localNetIPs ?

girishmg replied:
ok

trozet commented on go-controller/pkg/ovn/ovn.go:
```diff
@@ -76,8 +77,9 @@ type Controller struct {
-	joinSubnetAllocator   *allocator.SubnetAllocator
+	masterSubnetAllocator  *allocator.SubnetAllocator
+	joinSubnetAllocator    *allocator.SubnetAllocator
+	dnatAndSnatIPAllocator *ipallocator.Range
```
localNetIPAllocator

girishmg replied:
ok

trozet commented on go-controller/pkg/util/subnet_annotations.go:
```diff
@@ -23,6 +23,10 @@ import (
 //       {
 //         "default": "100.64.2.0/29"
 //       }
+//     k8s.ovn.org/node-dnat-and-snat-ip: |
```
lets not add annotation with this name. node-localnet-ips

girishmg replied:
ok

trozet replied:
fix the comment to be the new annotation

girishmg replied:
ok

girishmg commented on go-controller/pkg/node/gateway_shared_intf_linux.go:
```diff
@@ -1,17 +1,78 @@
+		return err
+	}
+
+	macAddress := util.IPAddrToHWAddr(net.ParseIP(util.V4DnatAndSnatNextHop)).String()
```
There is a difference in how these two subnets are being used. The semantics are different as well.

1. The subnet used for local gateway mode is per-node. Strictly speaking it only needs two IPs. So, it could be just 169.254.33.0/30. All we need is 169.254.33.1 and 169.254.33.2.

2. The subnet size for the logical switch that the distributed gateway port connects to is large. I have used /20, and with that prefix length we can support upt to 2^12 (4000+) nodes K8s cluster.

So, we can't merge those two subnets.

trozet replied:
Do either of these work?

1. set the values based on which mode is used OR
2. not set these values in util package and keep them local to the node package you are using (like gw local does)

girishmg replied:
@trozet the reason it is in util package is because I need those values from 
1. 'pkg/ovn' - the files under this package sets up the logical constructs using these values in ovnkube-master daemon
2. 'pkg/node' - the files under this package sets up the physical constructs (br-local, ovn-k8s-gw0, and so on) using these values in ovnkube-node daemon

girishmg commented on go-controller/pkg/ovn/gateway_cleanup.go:
```diff
@@ -41,10 +41,6 @@ func gatewayCleanup(nodeName string, nodeSubnets []*net.IPNet) error {
 		}
 	}
 
-	for _, nodeSubnet := range nodeSubnets {
```
That is correct

girishmg commented on go-controller/pkg/ovn/gateway.go:
```diff
@@ -20,6 +20,11 @@ const (
 	externalSwitchPrefix = "ext_"
 	gwRouterPrefix       = "GR_"
+
+	distributedGWPortSwitch = "DGP_ls"
```
@trozet Agree.  How about 
- `node_local_services_switch` or `nodelocal_services_switch` or `nodelocal_svcs_switch`
- `local_node_services_switch` or `localnode_services_switch` or `localnode_svcs_switch`
- `nodelocal_switch` or `localnode_switch`

trozet replied:
node_local_switch I think makes sense

girishmg commented on go-controller/pkg/node/gateway_localnet.go:
```diff
@@ -356,17 +357,24 @@ func localnetNodePortWatcher(ipt util.IPTablesHelper, wf *factory.WatchFactory,
+		m := strings.Split(bridgeMapping, ":")
+		if physnet == m[0] {
+			bridgeName := m[1]
+			_, stderr, err = util.RunOVSVsctl("--", "--if-exists", "del-br", bridgeName)
```
We don't want to delete all the OVS bridges. In the shared gateway mode, the host OVS bridge is responsible for the host stack, so we don't want to delete that OVS bridge. If we do, then the host network  will be hosted.

```
+------------------------------------------+
|Node1                                     |
|(10.8.51.51)                              |
|                                          |
|            +----------+                  |
|            | br-local |--------+         |
|            +----------+        |         |
|                                |         |
|                                |         |
|                       SNAT: NodeLocalIP  |
|                       +----------------+ |
|                       |     br-int     | |
|  +--------+           +------+---------+ |
|  |  Node  |             SNAT: NodeIP     |
|  |   IP   |                  |           |
|  +--------+                  |           |
|+-----------------------+     |           |
||        breth0         |   patch         |
||   NodePort OpenFlow   +-- cable         |
||         Flows         |                 |
|+-----------------------+                 |
|        +------+                          |
+--------+ eth0 +--------------------------+
         +------+                           
```
We don't want to delete `breth0` here.

trozet replied:
ah forgot about the extra bridge in shared gw mode..

girishmg commented on go-controller/pkg/node/gateway_shared_intf_linux.go:
```diff
@@ -1,17 +1,78 @@
 	"github.com/coreos/go-iptables/iptables"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
-
```
ok

girishmg commented on go-controller/pkg/ovn/gateway_cleanup.go:
```diff
@@ -97,9 +93,59 @@ func gatewayCleanup(nodeName string, nodeSubnets []*net.IPNet) error {
+	// delete the dnat_and_snat rule that we added for the management port IP
+	// Note: we don't need to delete any MAC bindings that are dynamically learned from OVN SB DB
+	// because there will be none since this NAT is only for outbound traffic and not for inbound
+	mgmtPortName := "k8s-" + nodeName
```
ok

girishmg commented on go-controller/pkg/ovn/gateway_cleanup.go:
```diff
@@ -97,9 +93,59 @@ func gatewayCleanup(nodeName string, nodeSubnets []*net.IPNet) error {
+		matches = ""
+	}
+	nodeSubnetMatchSubStr := fmt.Sprintf("rtos-%s", nodeName)
+	for _, match := range strings.Split(matches, "\n\n") {
```
all i need is one column, and using the default format is simple for this usecase.

girishmg commented on go-controller/pkg/ovn/gateway_cleanup.go:
```diff
@@ -97,9 +93,59 @@ func gatewayCleanup(nodeName string, nodeSubnets []*net.IPNet) error {
+		} else if strings.Contains(match, nodeName) {
+			priority = mgmtPortPolicyPriority
+		} else {
+			continue
```
that shouldn't happen because we are the ones adding the rules and the only rules allowed are `nodeSubnetPolicyPriority` and `mgmtPortPolicyPriority`, but i will log it.

girishmg replied:
Also, I am walking through all the PBR rules in the cluster to clear up the PBR rules for the node that is being removed from the cluster. So, I will find rules that don't match the current node, so I cannot  print them in the `else` branch above.

girishmg commented on go-controller/pkg/ovn/gateway_init.go:
```diff
@@ -304,6 +305,184 @@ func gatewayInit(nodeName string, clusterIPSubnet []*net.IPNet, hostSubnets []*n
+	}
+
+	// add a distributed gateway port to the distributed router
+	dgpName := "rtol-" + distributedGWPortSwitch
```
rtol  --> router-to-local, will provide information

girishmg replied:
@trozet just decided to use the existing, well-known prefix of `rtos-` and `stor-`

girishmg commented on go-controller/pkg/ovn/gateway_init.go:
```diff
@@ -304,6 +305,184 @@ func gatewayInit(nodeName string, clusterIPSubnet []*net.IPNet, hostSubnets []*n
+	}
+	// set gateway chassis (the current master node) for distributed gateway port)
+	nbctlArgs = append(nbctlArgs,
+		"--", "--id=@gw", "create", "gateway_chassis", "chassis_name="+masterChassisID, "external_ids:dgp_name="+dgpName,
```
That is exactly correct.

girishmg commented on go-controller/pkg/ovn/gateway_init.go:
```diff
@@ -304,6 +305,184 @@ func gatewayInit(nodeName string, clusterIPSubnet []*net.IPNet, hostSubnets []*n
+		return fmt.Errorf("failed creating logical switch %s and its ports (%s, %s) "+
+			"stdout: %q, stderr: %q, error: %v", distributedGWPortSwitch, lclNetPortname, lspName, stdout, stderr, err)
+	}
+	// finally add an entry to the OVN SB MAC_Binding table, if not present, to capture the
```
yes, we will need to file RFE for this against OVN.

trozet replied:
https://bugzilla.redhat.com/show_bug.cgi?id=1845111

girishmg replied:
Thanks

girishmg commented on go-controller/pkg/ovn/gateway_init.go:
```diff
@@ -304,6 +305,184 @@ func gatewayInit(nodeName string, clusterIPSubnet []*net.IPNet, hostSubnets []*n
+	return nil
+}
+
+func addPolicyBasedRoutes(nodeName, mgmtPortIP string, hostIfAddrs []*net.IPNet) error {
```
understood. Will fix it.

girishmg commented on go-controller/pkg/ovn/gateway_init.go:
```diff
@@ -304,6 +305,184 @@ func gatewayInit(nodeName string, clusterIPSubnet []*net.IPNet, hostSubnets []*n
+func addPolicyBasedRoutes(nodeName, mgmtPortIP string, hostIfAddrs []*net.IPNet) error {
+	// embed nodeName as comment so that it is easier to delete these rules later on.
+	// logical router policy doesn't support external_ids to stash metadata
+	matchStr := fmt.Sprintf(`inport == "rtos-%s" && ip4.dst == %s /* %s */`,
```
ok

girishmg commented on go-controller/pkg/ovn/gateway_init.go:
```diff
@@ -304,6 +305,184 @@ func gatewayInit(nodeName string, clusterIPSubnet []*net.IPNet, hostSubnets []*n
+func addPolicyBasedRoutes(nodeName, mgmtPortIP string, hostIfAddrs []*net.IPNet) error {
+	// embed nodeName as comment so that it is easier to delete these rules later on.
+	// logical router policy doesn't support external_ids to stash metadata
+	matchStr := fmt.Sprintf(`inport == "rtos-%s" && ip4.dst == %s /* %s */`,
```
@trozet the current comment in place is used as a metadata for me to delete the rule later. The `routing_policy` didn't have external_ids column till few months ago. So, I had to resort to using the comments as a means to delete the rule later.

girishmg commented on go-controller/pkg/ovn/gateway_init.go:
```diff
@@ -304,6 +305,184 @@ func gatewayInit(nodeName string, clusterIPSubnet []*net.IPNet, hostSubnets []*n
+		}
+	}
+
+	matchStr = fmt.Sprintf("ip4.src == %s && ip4.dst == %s /* %s */",
```
The comment is only used as a metadata because router_policy table didn't have external_ids column and it is only present in the master branch of OVN repository.

girishmg commented on go-controller/pkg/ovn/gateway_init.go:
```diff
@@ -304,6 +305,184 @@ func gatewayInit(nodeName string, clusterIPSubnet []*net.IPNet, hostSubnets []*n
+	return nil
+}
+
+func (oc *Controller) addPbrAndNatRules(node *kapi.Node, hostIfAddrs []*net.IPNet) error {
```
ok

trozet commented:
@nerdalert I see:

Jun  3 16:08:10.714: INFO: stdout: "PING 10.249.2.1 (10.249.2.1): 56 data bytes\n64 bytes from 10.249.2.1: seq=8 ttl=64 time=1.021 ms\n64 bytes from 10.249.2.1: seq=9 ttl=64 time=0.616 ms\n64 bytes from 10.249.2.1: seq=10 ttl=64 time=0.146 ms\n\n--- 10.249.2.1 ping statistics ---\n11 packets transmitted, 3 packets received, 72% packet loss\nround-trip min/avg/max = 0.146/0.594/1.021 ms\n"

The 2nd ping to the gw says 72% packet loss, 3 packets received, but I thought we only send 3 packets?

girishmg commented on go-controller/pkg/ovn/gateway_init.go:
```diff
@@ -304,6 +305,184 @@ func gatewayInit(nodeName string, clusterIPSubnet []*net.IPNet, hostSubnets []*n
+	_, stderr, err := util.RunOVNNbctl("lr-policy-add", ovnClusterRouter, nodeSubnetPolicyPriority, matchStr, "reroute",
+		mgmtPortIP)
+	if err != nil {
+		// lr-policy-add doesn't support --may-exist and what the heck?
```
@trozet filed https://github.com/ovn-org/ovn/issues/49, FYI

girishmg commented on go-controller/pkg/ovn/gateway_init.go:
```diff
@@ -304,6 +305,184 @@ func gatewayInit(nodeName string, clusterIPSubnet []*net.IPNet, hostSubnets []*n
+	return nil
+}
+
+func (oc *Controller) addDnatAndSnatRules(node *kapi.Node, mgmtPortMAC, mgmtPortIP string) error {
```
changed to addNodeLocalNATEntries

girishmg commented on go-controller/pkg/ovn/gateway_init.go:
```diff
@@ -304,6 +305,184 @@ func gatewayInit(nodeName string, clusterIPSubnet []*net.IPNet, hostSubnets []*n
+		annotationPresent = true
+		externalIP = externalIPs[0]
+	} else {
+		externalIP, err = oc.dnatAndSnatIPAllocator.AllocateNext()
```
done

girishmg commented on go-controller/pkg/ovn/gateway_init.go:
```diff
@@ -304,6 +305,184 @@ func gatewayInit(nodeName string, clusterIPSubnet []*net.IPNet, hostSubnets []*n
+	var externalIP net.IP
+
+	annotationPresent := false
+	externalIPs, err := util.ParseNodeDnatAndSnatIPAnnotation(node)
```
done

girishmg commented:
> Functionally it looks good. Would like a lot of the wording and code definitions changed. It would be great if you can also add testify unit tests for the new functions you are adding in addition to the ovn-nbctl tests for expected cmds.

@trozet I have addressed all of your comments except for the ones below..

1. I would like to tackle adding constants for prefixes `rtos-`, `stor-`, and `k8s-` in a separate PR since those were not newly added by this PR.
2. I have already added the BDD tests for the new code using fakeExec and gingko. The testify unit tests is something new to me, so I need to understand it first and then add it. So, I think that should not be a blocker for this PR since I already have BDD tests.

trozet commented on go-controller/pkg/ovn/gateway_init.go:
```diff
@@ -304,6 +305,168 @@ func gatewayInit(nodeName string, clusterIPSubnet []*net.IPNet, hostSubnets []*n
+			"stderr: %q, error: %v", util.V4NodeLocalNatSubnetNextHop, dnatSnatNextHopMac, dgpName, stderr, err)
+	}
+	if stdout != "" {
+		klog.Infof("The MAC_Binding entry of (%s, %s) exitsts on distributed router port %s with uuid %s",
```
s/exitsts/exists

girishmg replied:
Fixed

trozet commented on go-controller/pkg/util/subnet_annotations.go:
```diff
@@ -155,3 +162,63 @@ func SetNodeJoinSubnetAnnotation(nodeAnnotator kube.Annotator, defaultSubnets []
 	return parseSubnetAnnotation(node, ovnNodeJoinSubnets)
 }
+
+// CreateNodeLocalNatAnnotation creates a "k8s.ovn.org/node-dnat-and-snat-ip" annotation,
```
fix the comment for the annotation name

girishmg replied:
done

trozet commented on go-controller/pkg/util/subnet_annotations.go:
```diff
@@ -155,3 +162,63 @@ func SetNodeJoinSubnetAnnotation(nodeAnnotator kube.Annotator, defaultSubnets []
+	}, nil
+}
+
+// SetNodeLocalNatAnnotation sets a "k8s.ovn.org/node-dnat-and-snat-ip" annotation
```
ditto

girishmg replied:
done

trozet commented:
1 failure in np shared: [Fail] [sig-network] NetworkPolicy [LinuxOnly] NetworkPolicy between server and client [BeforeEach] should enforce multiple egress policies with egress allow-all policy taking precedence [Feature:NetworkPolicy]
